feat(cpn): support more than just FrSky S.Port telemetry simulation

nrw505 opened the pull request:
Fixes #4983
Fixes #2156 

Summary of changes:

1. Replace the current telemetry sensor UI dropdowns to choose a telemetry source for both the internal and external module.
2. Refactor the current FrSky S.port telemetry implementation to allow it to have separate telemetry for both internal and external modules at the same time
3. Update the simulator interface to accept different telemetry protocols
4. Implement a CRSF telemetry source and associated UI
5. Rejigger the way the log replay controller works so that it can work with multiple kinds of telemetry

Example: [Screencast from 10-08-24 01:17:55.webm](https://github.com/user-attachments/assets/fadfb3a1-dae1-4305-aa94-da30b1b10c8d)

elecpower commented:
Post merge of #5381 sim telemetry will need to be modified to suit surface radios eg flight vs drive and hide flight sensors eg vario

elecpower commented on companion/src/simulation/simulatedgps.h:
```diff
@@ -0,0 +1,72 @@
+ */
+
+#ifndef _SIMULATEDGPS_H_
+#define _SIMULATEDGPS_H_
```
Please use #pragma once for header guards throughout. It is a requirement for all new code and is progressively being performed as housekeeping by the dev team unless it does not achieve the desired compile code.

elecpower commented on companion/src/simulation/telemetryprovider.h:
```diff
@@ -0,0 +1,50 @@
+ */
+
+#ifndef _TELEMETRYPROVIDER_H_
+#define _TELEMETRYPROVIDER_H_
```
Fix required

elecpower commented on companion/src/simulation/telemetryprovidercrossfire.cpp:
```diff
@@ -0,0 +1,530 @@
+          ui->input_gps,  &QLineEdit::setText);
+  connect(&gps,           &SimulatedGPS::courseDegreesChanged,
+          [this](double v) { ui->input_hdg->setValue(v); });
+
```
For code consistency with existing code please keep Qt::connect functions to a single line. We usually only split across lines when lambda functions used for the receiver.
Please replace the use of SIGNAL and SLOT macros with function pointers and lambda functions. The macros do not provide compile time checking of the sender and receiver events.

elecpower commented on companion/src/simulation/telemetryprovidercrossfire.h:
```diff
@@ -0,0 +1,77 @@
+ */
+
+#ifndef _TELEMETRY_CROSSFIRE_H_
+#define _TELEMETRY_CROSSFIRE_H_
```
Amend

elecpower commented on companion/src/simulation/telemetryproviderfrsky.h:
```diff
@@ -0,0 +1,114 @@
+ */
+
+#ifndef _TELEMETRY_FRSKY_H_
+#define _TELEMETRY_FRSKY_H_
```
Amend

elecpower commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -76,7 +67,16 @@ TelemetrySimulator::TelemetrySimulator(QWidget * parent, SimulatorInterface * si
+          simulator,           &SimulatorInterface::sendInternalModuleTelemetry);
+  connect(this,                &TelemetrySimulator::externalTelemetryDataChanged,
+          simulator,           &SimulatorInterface::sendExternalModuleTelemetry);
+
```
Amend

elecpower commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
+TelemetryProvider * TelemetrySimulator::newTelemetryProviderFromDropdownChoice(const QString &text, QScrollArea * parent, bool isExternal)
 {
-  logPlayback->stepForward(false);
+  if (text == "None") {
```
This an subsequent branch tests will not achieve the desired result when the text is translated in the drop down list. Use the currentIndex instead

elecpower commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
+  if (text == "None") {
+    return NULL;
+  }
+  if (text == "FrSky S.Port") {
```
amend

elecpower commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
+    }
+    return newProvider;
+  }
+  if (text == "CRSF / ELRS") {
```
amend

elecpower commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
+              this,        &TelemetrySimulator::onExternalTelemetryProviderDataChanged);
+    } else {
+      connect(newProvider, &TelemetryProviderFrSky::telemetryDataChanged,
+              this,        &TelemetrySimulator::onInternalTelemetryProviderDataChanged);
```
amend connects

elecpower commented on companion/src/simulation/telemetrysimu.cpp:
```diff
@@ -153,53 +161,78 @@ void TelemetrySimulator::onSimulateToggled(bool isChecked)
+              this,        &TelemetrySimulator::onExternalTelemetryProviderDataChanged);
+    } else {
+      connect(newProvider, &TelemetryProviderCrossfire::telemetryDataChanged,
+              this,        &TelemetrySimulator::onInternalTelemetryProviderDataChanged);
```
amend

elecpower commented on companion/src/simulation/telemetrysimu.h:
```diff
@@ -28,8 +28,10 @@
 #include <QDateTime>
 #include <QtCore/qmath.h>
 #include <QFileDialog>
+#include <QScrollArea>
```
Header guard above

elecpower commented on companion/src/simulation/telemetryprovidercrossfire.cpp:
```diff
@@ -0,0 +1,530 @@
+  QString inputText;
+  double inputDouble;
+  int inputInt;
+
```
Currently possible to inadvertently load Frsky saved values. Needs a test and rejection msg to save the user from themselves

elecpower commented on companion/src/simulation/telemetryproviderfrsky.cpp:
```diff
@@ -0,0 +1,934 @@
+        return;
+    }
+
+    QTextStream in(&file);
```
Currently possible to inadvertently load crsf saved values. Needs a test and rejection msg to save the user from themselves

elecpower reviewed (changes requested):
UI comments:
- change buttons font size to default (just look out of place in Linux)
- QComboBox protocol selector would look better IMO if moved closer to label

nrw505 commented:
> * change buttons font size to default (just look out of place in Linux)

They already are the default font size, the .ui file does not specify a font size for the buttons e.g.

```
    <widget class="QPushButton" name="button_saveTelemetryValues">
     <property name="text">
      <string>Save Telemetry Values</string>
     </property>
    </widget>
```

This is the same font size that the buttons are in the current frsky-only telemetry screen (and the fields and labels are 8pt as they are on the current frsky-only telemetry screen).

I'm happy to change them to 8 point so they match the rest of the fields on that widget it that's what you want. Please confirm.

> * QComboBox protocol selector would look better IMO if moved closer to label

Just right-align the labels so they're next to the QComboBox OK?

elecpower commented:
> This is the same font size that the buttons are in the current frsky-only telemetry screen (and the fields and labels are 8pt as they are on the current frsky-only telemetry screen).
> I'm happy to change them to 8 point so they match the rest of the fields on that widget it that's what you want. Please confirm.

Make buttons consistent with rest of dialog.

> Just right-align the labels so they're next to the QComboBox OK?

Right align would suffice.

IMO the dialog could do with some grids to aid in layout but that is for another day.

LupusTheCanine commented:
whould it be possible to use serial port as telemetry source and possibly output port for protocols like CRSF?

nrw505 commented:
> whould it be possible to use serial port as telemetry source and possibly output port for protocols like CRSF?

If you wanted to do that I think you'd probably be better off changing radio/src/targets/simu to use a real serial port as an internal (or external) module - then you'd have _real_ telemetry and the simulation telemetry wouldn't be used at all.

pfeerick commented:
My only nitpicks so far would be:
- for CRSF/ELRS, FM has quotes which shouldn't be there
- 1RSS isn't being read from the attached log file (TX16S, ELRS, Betaflight).  [TinyTrainer2-2024-08-03-091029.zip](https://github.com/user-attachments/files/16644282/TinyTrainer2-2024-08-03-091029.zip)
- Could do with some padding at the top for the checkboxes... 
![image](https://github.com/user-attachments/assets/a8b2cf9a-821d-400c-94b1-efbe8aee6a12)

Otherwise, after a quick poke around and replay of the log... this is working fantastic! :)

pfeerick commented:
btw, don't think I didn't notice this! 🤣 🤣 🤣 
![image](https://github.com/user-attachments/assets/99517b44-ed15-4960-bd04-6b1c77f6b2bd)

nrw505 commented:
> * for CRSF/ELRS, FM has quotes which shouldn't be there

Fixed

> * 1RSS isn't being read from the attached log file (TX16S, ELRS, Betaflight).  [TinyTrainer2-2024-08-03-091029.zip](https://github.com/user-attachments/files/16644282/TinyTrainer2-2024-08-03-091029.zip)

It's being read OK, but I had the UI set up wrong given CSRF RSS measurements are an actual measurement not a percentage-like value like S.Port. Fixed.

> * Could do with some padding at the top for the checkboxes... (ignore this if already covered earlier)

Oops, didn't mean to change that part of the original UI. Padding added.

nrw505 commented:
> btw, don't think I didn't notice this! 🤣 🤣 🤣 ![image](https://private-user-images.githubusercontent.com/5500713/358838632-99517b44-ed15-4960-bd04-6b1c77f6b2bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM4OTA2ODIsIm5iZiI6MTcyMzg5MDM4MiwicGF0aCI6Ii81NTAwNzEzLzM1ODgzODYzMi05OTUxN2I0NC1lZDE1LTQ5NjAtYmQwNC02YjFjNzdmNmIyYmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTdUMTAyNjIyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MGFiNDBlNGZlNTBkMWVkZWIxOGNkZWEyNGY1OGI5MjFhOThmMzAwOGY0Y2E0ZDUzYmMyOGFjODk1NTE4ZDY0NyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.mkTXp42pZrygU-hXXltHh1T2RFhUmlBIQkQAEMWi0h0)

Hey, you said you wanted it merge-ready :P . Keep an eye out for followup PRs. I also want to change what's selectable in the internal/external module dropdowns based on what the simulated radio is capable of and maybe a warning if you choose something that the current model isn't configured for.

3djc commented:
Looks quite nice to me. 

Unsure if in the scope of this contest, but I think ELRS has this specificity that you can set frame rate and telemetry frame ratio. When set to very low frame rate (like 25hz) and low telemetry ratio, the flow of incoming telemetry is quite slow, and may potentially affect Lua scripts. It would be great if (even at a later time) you could have a way to specify 'I'm running at 25Hz, 1/32 telem' (or whatever those values are)

But already quite nice !

pfeerick commented:
That's probably out of scope, but would be interesting to see. Do you see that primarily only affecting the replay of a log? Thus effectively just more granularity over the replay rate? Or are you after it slowing down the refreshes of the sensors, i.e. for ELRS to be taking into account the [LINK and DATA groupings and data rates for sensors](https://www.expresslrs.org/info/telem-bandwidth/).

3djc commented:
I was thinking 'live' usage, affecting the refresh of sensors, for LUA devs who might make computation based on sensor that might glitch at low or high refresh speed

nrw505 commented:
Oops wrong button sorry

3djc commented:
I was not thinking log at all, only live data, that could have a "jitter" value of 0 (so like today) or say x% where value would vary by that much, as it is currently difficult when you debug scripts on simu to see how your values follow (or not) sensor value changes (but then again , out of scope of the competition itself)

nrw505 commented:
@3djc you mean kind of like how the simulated GPS can currently be set to run its own little simulation which updates the position over time?

3djc commented:
> @3djc you mean kind of like how the simulated GPS can currently be set to run its own little simulation which updates the position over time?

Something like that
![image](https://github.com/user-attachments/assets/e2dcc848-9318-41bf-9ecd-810bede0515c)

where you could decide an amount of variation. If you wanted to go over the top, you would add another column to decide variation type (rng, sine, triangle, square,..), but a default of triangle or rng would  be perfect

nrw505 commented:
Rebased to squash all the fixups

pfeerick commented:
Probably my final comments on this before merge:
- can we make it so don't need to re-enter instance ids for the FrSky SPort fields before they function? As now need to reenter "something" before can replay frsky telemetry.
- from a BF ELRS quad, GPS lat long appears to not be comma separated, so doesn't relay through to simulation from the logs (i.e. see [Crux3 NLR-2022-10-08-093345.csv.zip](https://github.com/user-attachments/files/16667282/Crux3.NLR-2022-10-08-093345.csv.zip))...

nrw505 commented:
> * can we make it so don't need to re-enter instance ids for the FrSky SPort fields before they function? As now need to reenter at least "something" before can replay frsky telemetry.

Whoops, had some initialisation for the FrSky provider called by the wrong spot: now it populates those instance id fields like it did before this PR.

> * from a BF ELRS quad, GPS lat long appears to not be comma separated, so doesn't relay through to simulation from the logs

Made the simulated GPS accept space separated lat/long as well as tolerating spaces around a comma.
